src/deviceApiCalls/transports/android.transport.js

```diff
@@ -0,0 +1,117 @@
+import {DeviceApiTransport} from '../../../packages/device-api'
+import {
+    GetAutofillDataCall,
+    GetAvailableInputTypesCall,
+    GetRuntimeConfigurationCall,
+    StoreFormDataCall
+} from '../__generated__/deviceApiCalls'
+
+export class AndroidTransport extends DeviceApiTransport {
+    /** @type {GlobalConfig} */
+    config
+
+    /** @param {GlobalConfig} globalConfig */
+    constructor (globalConfig) {
+        super()
+        this.config = globalConfig
+    }
+    /**
+     * @param {import("../../../packages/device-api").DeviceApiCall} deviceApiCall
+     * @returns {Promise<any>}
+     */
+    async send (deviceApiCall) {
+        if (deviceApiCall instanceof GetRuntimeConfigurationCall) {
+            return androidSpecificRuntimeConfiguration(this.config)
+        }
+
+        if (deviceApiCall instanceof GetAvailableInputTypesCall) {
+            return androidSpecificAvailableInputTypes(this.config)
+        }
+
+        if (deviceApiCall instanceof GetAutofillDataCall) {
+            window.BrowserAutofill.getAutofillData(JSON.stringify(deviceApiCall.params))
+            return waitForResponse(deviceApiCall.id, this.config)
+        }
+        if (deviceApiCall instanceof StoreFormDataCall) {
+            return window.BrowserAutofill.storeFormData(JSON.stringify(deviceApiCall.params))
+        }
+        throw new Error('android: not implemented: ' + deviceApiCall.method)
+    }
+}
+
+/**
+ * Sends a message and returns a Promise that resolves with the response
+ *
+ * @param {string} expectedResponse - the name of the response
+ * @param {GlobalConfig} config
+ * @returns {Promise<*>}
+ */
+function waitForResponse (expectedResponse, config) {
+    return new Promise((resolve) => {
+        const handler = e => {
+            if (!config.isDDGTestMode) {
+                if (e.origin !== '') {
```

Comment:
Mhm, interesting. So the messages we send from the native layer have `origin = ''`?

Reply:
Yep that's correct - this is something we noticed (it's similar on windows). It appears that no page script could ever construct such an event, so this is a nice way of us filtering messages to those originating from the app.

Reply:
Nice!